src/transform/src/predicate_propagation.rs

```diff
@@ -0,0 +1,837 @@
+                            if let expr::AggregateFunc::Count = aggregate.func {
+                                // Replace counts of nonnull, nondistinct expressions with `true`.
+                                if let MirScalarExpr::Column(c) = aggregate.expr {
+                                    if !input_type.column_types[c].nullable && !aggregate.distinct {
+                                        aggregate.expr = MirScalarExpr::literal_ok(
+                                            Datum::True,
+                                            ScalarType::Bool,
+                                        );
+                                    }
+                                }
+                                predicates.push(Constraint::Predicate(
+                                    MirScalarExpr::column(column)
+                                        .call_unary(UnaryFunc::IsNull(expr::func::IsNull))
+                                        .call_unary(UnaryFunc::Not(expr::func::Not)),
+                                ));
+                            } else {
+                                // TODO: Something more sophisticated using `input_knowledge` too.
+                            }
+                        }
+                    }
+
+                    predicates
+                }
+                MirRelationExpr::TopK { input, .. } => self.action(input, let_knowledge)?,
+                MirRelationExpr::Negate { input } => self.action(input, let_knowledge)?,
+                MirRelationExpr::Threshold { input } => self.action(input, let_knowledge)?,
+                MirRelationExpr::Union { base, inputs } => {
+                    let mut know1 = denormalize_predicates(self.action(base, let_knowledge)?);
+                    for input in inputs.iter_mut() {
+                        let know2 = denormalize_predicates(self.action(input, let_knowledge)?);
+                        know1.retain(|predicate| know2.contains(predicate));
+                    }
+                    know1
+                }
+            };
+
+            // Propagate the nullability flags of the projected columns as predicates
+            predicates.extend(self_type.column_types.iter().enumerate().flat_map(
+                |(col_idx, col_typ)| {
+                    if !col_typ.nullable {
+                        Some(Constraint::Predicate(
+                            MirScalarExpr::column(col_idx)
+                                .call_unary(UnaryFunc::IsNull(expr::func::IsNull))
+                                .call_unary(UnaryFunc::Not(expr::func::Not)),
+                        ))
+                    } else {
+                        None
+                    }
+                },
+            ));
+
+            normalize_predicates(&mut predicates, &self_type);
+            if predicates
+                .iter()
+                .any(|p| p.is_literal_false() || p.is_literal_null())
+            {
+                expr.take_safely();
```

Comment:
`normalize_predicates` already removes `true` predicates.

We can end up with other literal predicates here because `normalize_predicates` calls `reduce` which may convert an existing predicate into a literal.